src/localplayer.cpp

```diff
@@ -639,86 +640,90 @@ void LocalPlayer::applyControl(float dtime)
+	} else {
+		if (control.up) {
+			if (continuous_forward) {
 				if (fast_move)
-					speedV.Y = movement_speed_fast;
-				else
-					speedV.Y = movement_speed_walk;
+					superspeed = true;
 			} else {
-				if(fast_move && control.aux1)
-					speedV.Y = movement_speed_fast;
-				else
-					speedV.Y = movement_speed_walk;
+				speedH += move_direction;
 			}
 		}
-		else if(m_can_jump)
-		{
-			/*
-				NOTE: The d value in move() affects jump height by
-				raising the height at which the jump speed is kept
-				at its starting value
-			*/
-			v3f speedJ = getSpeed();
-			if(speedJ.Y >= -0.5 * BS) {
-				if (m_ledge_detected) {
-					// Limit jump speed to a minimum that allows
-					// jumping up onto a ledge 2 nodes up.
-					speedJ.Y = movement_speed_jump *
-							MYMAX(physics_override_jump, 1.3f);
-					setSpeed(speedJ);
-					m_ledge_detected = false;
+		if (control.down) {
+			speedH -= move_direction;
+		}
+		if (!control.up && !control.down) {
+			speedH -= move_direction *
+					  (control.forw_move_joystick_axis / 32767.f);
+		}
+		if (control.left) {
+			speedH += move_direction.crossProduct(v3f(0, 1, 0));
+		}
+		if (control.right) {
+			speedH += move_direction.crossProduct(v3f(0, -1, 0));
+		}
+		if (!control.left && !control.right) {
+			speedH -= move_direction.crossProduct(v3f(0, 1, 0)) *
+					  (control.sidew_move_joystick_axis / 32767.f);
+		}
+		if (control.jump) {
+			if (free_move) {
+				if (g_settings->getBool("aux1_descends") || always_fly_fast) {
+					if (fast_move)
+						speedV.Y = movement_speed_fast;
+					else
+						speedV.Y = movement_speed_walk;
 				} else {
-					speedJ.Y = movement_speed_jump * physics_override_jump;
-					setSpeed(speedJ);
+					if (fast_move && control.aux1)
+						speedV.Y = movement_speed_fast;
+					else
+						speedV.Y = movement_speed_walk;
 				}
+			} else if (m_can_jump) {
+				/*
+					NOTE: The d value in move() affects jump height by
+					raising the height at which the jump speed is kept
+					at its starting value
+				*/
+				v3f speedJ = getSpeed();
+				if (speedJ.Y >= -0.5 * BS) {
+					if (m_ledge_detected) {
+						// Limit jump speed to a minimum that allows
+						// jumping up onto a ledge 2 nodes up.
+						speedJ.Y = movement_speed_jump *
+								   MYMAX(physics_override_jump, 1.3f);
+						setSpeed(speedJ);
+						m_ledge_detected = false;
+					} else {
+						speedJ.Y = movement_speed_jump * physics_override_jump;
+						setSpeed(speedJ);
+					}
 
-				MtEvent *e = new SimpleTriggerEvent("PlayerJump");
-				m_client->event()->put(e);
+					MtEvent *e = new SimpleTriggerEvent("PlayerJump");
+					m_client->event()->put(e);
+				}
+			} else if (in_liquid) {
+				if (fast_climb)
+					speedV.Y = movement_speed_fast;
+				else
+					speedV.Y = movement_speed_walk;
+				swimming_vertical = true;
+			} else if (is_climbing) {
+				if (fast_climb)
+					speedV.Y = movement_speed_fast;
+				else
+					speedV.Y = movement_speed_climb;
 			}
 		}
```

Comment:
No need to block other inputs, when a player is in mid-jump or is falling, they can still move, this should happen during knockback too.